### PR TITLE
Remove implicit clone

### LINKED ISSUE #254: Remove implicit clone in stype encoders
Currently, we clone the `TensorData` in `StypeEncoders` to avoid in place modifications. But this is bad because it is wasteful. We want to remove that clone while avoiding inplace modification.

### PR DESCRIPTION
**Pull request by yiweny**
Fixes #254 

Summary:
1. Only clone when `na_strategy` is specified in stype encoder. Only clone the `values` of a `MultiTensor`.

### CODE REVIEW

File: test/nn/encoder/test_stype_encoder.py [hunk 1]
```diff
@@ -22,13 +26,29 @@
...
     TimestampEncoder,
 )
-from torch_frame.testing.text_embedder import HashTextEmbedder
 from torch_frame.testing.text_tokenizer import (
     RandomTextModel,
     WhiteSpaceHashTokenizer,
 )
 
 
+def eq(td1: TensorData, td2: TensorData):
```
Comment (rusty1s) [inline test/nn/encoder/test_stype_encoder.py:35]: We can move this function to the package directly.

File: torch_frame/data/multi_nested_tensor.py [hunk 2]
```diff
@@ -65,6 +65,13 @@ def validate(self):
         assert self.offset[-1] == len(self.values)
         assert len(self.offset) == self.num_rows * self.num_cols + 1
 
+    def __eq__(self, td: object) -> bool:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:68]: FYI: This is a bit different from how PyTorch tensors deal with equality. If we want to keep consistency, `__eq__` would need to return a mask, and equality would need to be checked via `tensor.equal()`/`torch.eq` and `tensor.allclose()`.

File: torch_frame/data/multi_tensor.py [hunk 3]
```diff
@@ -325,6 +325,37 @@ def select(
     def _single_index_select(self, index: int, dim: int) -> _MultiTensor:
         raise NotImplementedError
 
+    def fillna(self, fill_value: Union[float, int]) -> None:
+        is_float = self.dtype.is_floating_point
+        if (is_float and isinstance(fill_value, int)
```
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:330]: I don't see an issue filling with an integer for floating-point values. Do we need this check? Why not let PyTorch handle any errors downstream.

File: torch_frame/data/multi_tensor.py [hunk 4]
```diff
@@ -325,6 +325,37 @@ def select(
     def _single_index_select(self, index: int, dim: int) -> _MultiTensor:
         raise NotImplementedError
 
+    def fillna(self, fill_value: Union[float, int]) -> None:
```
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:328]: Would be great if we can allow column-level vectors here as well.

File: torch_frame/data/multi_tensor.py [hunk 5]
```diff
@@ -325,6 +325,37 @@ def select(
...
+            fill_value: Any
+        """
+        self._fill_col(index, fill_value)
+
+    def _fill_col(
+        self,
+        index: int,
+        fill_value: Union[float, int],
+    ) -> None:
+        raise NotImplementedError
```
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:357]: Any reason we need `_fill_col` and `fill_col`?

File: torch_frame/__init__.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
     embedding,
 )
 from .data import TensorFrame
-from .typing import TaskType, Metric, DataFrame, NAStrategy
+from .typing import TaskType, Metric, DataFrame, NAStrategy, TensorData
```
Comment (rusty1s) [inline torch_frame/__init__.py:14]: Where is this helpful?

File: torch_frame/data/multi_embedding_tensor.py [hunk 7]
```diff
@@ -201,6 +208,19 @@ def _single_index_select(
             )
         assert False, "Should not reach here."
 
+    def _fill_col(
+        self,
+        index: int,
+        fill_value: Union[float, int],
+    ) -> None:
```
Comment (rusty1s) [inline torch_frame/data/multi_embedding_tensor.py:215]: Do you mean `fillna_col`?

File: torch_frame/data/multi_embedding_tensor.py [hunk 8]
```diff
@@ -201,6 +208,19 @@ def _single_index_select(
...
+        value_index = slice(self.offset[index], self.offset[index + 1])
+        values = self.values[:, value_index]
+        if self.values.is_floating_point():
+            values[torch.isnan(values)] = fill_value
+        else:
+            values[values == -1] = fill_value
+        self.values[:, value_index] = values
```
Comment (rusty1s) [inline torch_frame/data/multi_embedding_tensor.py:222]: Do we need this? Looks like everything before is zero-copy anyway.

File: torch_frame/nn/encoder/stype_encoder.py [hunk 9]
```diff
@@ -37,6 +37,19 @@ def reset_parameters_soft(module: Module):
         module.reset_parameters()
 
 
+def get_nan_mask(tensor: Tensor):
```
Comment (rusty1s) [inline torch_frame/nn/encoder/stype_encoder.py:40]:
```suggestion
def get_nan_mask(tensor: Tensor) -> Tensor:
```

File: torch_frame/nn/encoder/stype_encoder.py [hunk 10]
```diff
@@ -174,21 +182,11 @@ def na_forward(self, feat: TensorData) -> TensorData:
         """
         if self.na_strategy is None:
             return feat
-
+        if (isinstance(feat, _MultiTensor) and get_nan_mask(feat.values).any()
+                or isinstance(feat, Tensor) and get_nan_mask(feat).any()):
+            # Only clone the feat if there are NaNs in the data
+            feat = feat.clone()
```
Comment (rusty1s) [inline torch_frame/nn/encoder/stype_encoder.py:188]: It's probably more expensive to check this here than just doing the copy.

File: torch_frame/data/multi_tensor.py (hunk 4 above)
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:328]: I think it would help us to use PyTorch notation for in-place ops, e.g., in this case, this should be `fillna_`.

Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:328]:
```suggestion
    def fillna(self, fill_value: Union[float, int]) -> Self:
```

File: torch_frame/nn/encoder/stype_encoder.py (hunk 10 above)
Comment (rusty1s) [inline torch_frame/nn/encoder/stype_encoder.py:188]: minor nit: The `clone` here would also clone offsets, while we would only need to clone the values.

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (yiweny) [reply torch_frame/data/multi_nested_tensor.py:68]: I feel like returning a mask would be inefficient here. Because then we need to convert the two `MultiTensor`s to their dense format. The current implementation is faster. Or do you mean return a tuple of mask, one for the `values` and one for the `offset`?

File: torch_frame/data/multi_tensor.py (hunk 5 above)
Comment (yiweny) [reply torch_frame/data/multi_tensor.py:357]: Removed.

File: torch_frame/data/multi_tensor.py (hunk 4 above)
Comment (yiweny) [reply torch_frame/data/multi_tensor.py:328]: Got it.

Comment (yiweny) [reply torch_frame/data/multi_tensor.py:328]: I get self is not defined error if I change it to self.

File: torch_frame/__init__.py (hunk 6 above)
Comment (yiweny) [reply torch_frame/__init__.py:14]: In `eq` function used by tests, but we can move that into package.

File: torch_frame/data/multi_embedding_tensor.py (hunk 8 above)
Comment (yiweny) [reply torch_frame/data/multi_embedding_tensor.py:222]: Thanks. What you say make sense!

File: torch_frame/nn/encoder/stype_encoder.py (hunk 10 above)
Comment (yiweny) [reply torch_frame/nn/encoder/stype_encoder.py:188]: Changed the code to only clone the `values` tensor instead.

File: test/nn/encoder/test_stype_encoder.py [hunk 11]
```diff
@@ -324,40 +333,6 @@ def test_timestamp_feature_encoder_with_nan(encoder_cls_kwargs):
     assert (~torch.isnan(x)).all()
 
 
-def test_text_embedded_encoder():
```
Comment (yiweny) [inline test/nn/encoder/test_stype_encoder.py:327]: This test is no longer necessary because `text_embedded` is converted to `embedding` during materialization.

File: torch_frame/data/multi_nested_tensor.py [hunk 12]
```diff
@@ -275,6 +282,27 @@ def _single_index_select(self, index: int, dim: int) -> MultiNestedTensor:
         else:
             raise RuntimeError(f"Unsupported dim={dim} for index_select.")
 
+    def fillna_col(
+        self,
+        index: int,
+        fill_value: Union[int, float, Tensor],
+    ) -> None:
```
Comment (zechengz) [inline torch_frame/data/multi_nested_tensor.py:289]: Add docs that we inplace modify the tensor?

File: torch_frame/data/multi_tensor.py [hunk 13]
```diff
@@ -325,6 +324,49 @@ def select(
     def _single_index_select(self, index: int, dim: int) -> _MultiTensor:
         raise NotImplementedError
 
+    def fillna_(self, fill_value: Union[int, float, Tensor]):
+        """Fill the :obj:`MultiTensor` with fill_value.
```
Comment (zechengz) [inline torch_frame/data/multi_tensor.py:328]: Also, indicate this is an inplace operation?

File: torch_frame/nn/encoder/stype_encoder.py [hunk 14]
```diff
@@ -37,6 +37,19 @@ def reset_parameters_soft(module: Module):
...
+def get_nan_mask(tensor: Tensor) -> Tensor:
+    r"""Obtains the NaN maks of the input :obj:`Tensor`.
+
+    Args:
+        tensor (Tensor): Input :obj:`Tensor`.
+    """
+    if tensor.is_floating_point():
+        nan_mask = torch.isnan(tensor)
+    else:
+        nan_mask = tensor == -1
```
Comment (zechengz) [inline torch_frame/nn/encoder/stype_encoder.py:49]: This means na mask rather than nan mask for categorical column right?

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (zechengz) [reply torch_frame/data/multi_nested_tensor.py:68]: I think we should try to keep consistent with PyTorch functions.

File: test/data/test_multi_embedding_tensor.py [hunk 15]
```diff
@@ -99,6 +99,52 @@ def test_from_tensor_list():
     assert torch.allclose(met.offset, expected_offset)
     assert_equal(tensor_list, met)
 
+    met_with_nan = met.clone()
+    met_with_nan.values[0::2] = -1
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:103]: Does this mask the entire `met.values`?

File: test/data/test_multi_embedding_tensor.py [hunk 16]
```diff
@@ -99,6 +99,52 @@ def test_from_tensor_list():
...
+    met_4 = met_with_nan
+    met_4.fillna_(torch.tensor([[100, 101, 102]]))
+    assert torch.all(
+        torch.eq(
+            met_4.values,
+            torch.tensor([
+                [100, 100, 100, 101, 101, 102],
+                [3, 4, 5, 8, 9, 11],
+            ])))
+    assert torch.all(torch.eq(met_4.offset, expected_offset))
```
Comment (weihua916) [inline test/data/test_multi_embedding_tensor.py:138]: I don't understand these...

File: torch_frame/data/multi_nested_tensor.py [hunk 17]
```diff
@@ -65,6 +65,13 @@ def validate(self):
         assert self.offset[-1] == len(self.values)
         assert len(self.offset) == self.num_rows * self.num_cols + 1
 
+    def __eq__(self, td: object) -> bool:
+        if not isinstance(td, MultiNestedTensor):
+            raise NotImplementedError
+        return bool(
+            torch.all(torch.eq(self.offset, td.offset))
+            and torch.all(torch.eq(self.values, td.values)))
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:73]: `torch.eq` --> `torch.allclose`

File: torch_frame/data/multi_nested_tensor.py (hunk 12 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:289]: +1. I don't get the behavior even from the test.

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (rusty1s) [reply torch_frame/data/multi_nested_tensor.py:68]: I am just saying you should rename this function to `eq` and `equal` :)

**Review by weihua916 — COMMENTED** [review]
I think this PR can be much simpler. I am sorry but I don't understand why we are introducing fully-fledged fillna functionalities, while only a small fraction of it is really used for the main purpose of the PR.

Also, I personally had a different way of design. I am in favor of some redesign...

File: torch_frame/data/multi_tensor.py [hunk 18]
```diff
@@ -325,6 +324,49 @@ def select(
...
+
+        Args:
+            index (int): A column index of the tensor to select.
+            fill_value (Union[int, float, Tensor]): Scalars or vector
+                tensor of size (num_cols,) or (1, num_cols) to replace
+                NaNs.
+        """
+        if isinstance(fill_value, Tensor):
+            if fill_value.dim(
+            ) != 0 and not fill_value.numel() == self.num_cols:
```
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:338]: Do you mean `dim() != 1`?

Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:338]:
```suggestion
            if fill_value.dim(
            ) != 0 and fill_value.numel() != self.num_cols:
```

File: torch_frame/data/multi_tensor.py [hunk 19]
```diff
@@ -325,6 +324,49 @@ def select(
...
+                raise ValueError(
+                    "fillna_ with Tensor of shape "
+                    f"{fill_value.shape} is not supported. "
+                    f"Please use a Tensor of shape ({self.num_cols},)"
+                    f" or (1, {self.num_cols}).")
+            if fill_value.dim() != 0:
+                if fill_value.dim() == 1:
+                    fill_value = fill_value.unsqueeze(0)
```
Comment (rusty1s) [inline torch_frame/data/multi_tensor.py:346]: Not sure I get all this `dim` logic here. Why not just do `fill_value.view(-1)` at the start?

File: torch_frame/nn/encoder/stype_encoder.py [hunk 20]
```diff
@@ -208,7 +220,28 @@ def na_forward(self, feat: TensorData) -> TensorData:
...
+                feat.fillna_col(col, fill_value)
+            else:
+                column_data = feat[:, col]
+                nan_mask = get_nan_mask(column_data)
+                if nan_mask.ndim == 2:
+                    nan_mask = nan_mask.any(dim=-1)
+                assert nan_mask.ndim == 1
+                assert len(nan_mask) == len(column_data)
+                if not nan_mask.any():
+                    continue
```
Comment (rusty1s) [inline torch_frame/nn/encoder/stype_encoder.py:233]:
```suggestion
```

File: test/nn/encoder/test_stype_encoder.py [hunk 21]
```diff
@@ -44,10 +45,12 @@ def test_categorical_feature_encoder(encoder_cls_kwargs):
         stype=stype.categorical,
         **encoder_cls_kwargs[1],
     )
-    feat_cat = tensor_frame.feat_dict[stype.categorical]
+    feat_cat = tensor_frame.feat_dict[stype.categorical].clone()
```
Comment (weihua916) [inline test/nn/encoder/test_stype_encoder.py:48]: This does not really test anything since the `dataset` does not contain `nan` in the first place. Same for others.

Comment (yiweny) [reply test/nn/encoder/test_stype_encoder.py:48]: I am testing the in-place modification here. So `feat_cat` should be the same value has the `tensor_frame.feat_dict[stype.categorical]` after being passed to encoders

Comment (weihua916) [reply test/nn/encoder/test_stype_encoder.py:48]: Yes I understand. My point is that the original data does not contain any NaN, so there is likely no in-place modification at all. What's interesting to test is when the data contains NaN.

**Comment by yiweny** [comment]
There are some confusions regarding this PR that we resolved offline. Basically, @weihua916 didn't realize that the pr is also trying to solve a bug regarding multicategorical nan masking #301 . I will make it more clear by separating out this pr into two. One to remove implicit clone and another to fix nan masking. This will target 0.2.1 release.

File: test/nn/encoder/test_stype_encoder.py (hunk 21 above)
Comment (yiweny) [reply test/nn/encoder/test_stype_encoder.py:47]: Clarified offline that this is testing there there are no-inplace modifications without nans

File: test/nn/encoder/test_stype_encoder.py [hunk 22]
```diff
@@ -421,11 +406,23 @@ def test_text_tokenized_encoder():
         stype=stype.text_tokenized,
         col_to_model_cfg=col_to_model_cfg,
     )
-    feat_text = tensor_frame.feat_dict[stype.text_tokenized]
+    feat_text = copy.deepcopy(tensor_frame.feat_dict[stype.text_tokenized])
```
Comment (rusty1s) [inline test/nn/encoder/test_stype_encoder.py:409]: Does `text_tokenized` do not use `MultiNestedTensor` yet?

**Review by rusty1s — APPROVED** [review]

File: test/nn/encoder/test_stype_encoder.py (hunk 22 above)
Comment (yiweny) [reply test/nn/encoder/test_stype_encoder.py:409]: `text_tokenized` use a dictionary of `MultiEmbeddingTensor`, so `deepcopy` is necessary here.

File: torch_frame/nn/encoder/stype_encoder.py [hunk 23]
```diff
@@ -174,20 +182,36 @@ def na_forward(self, feat: TensorData) -> TensorData:
...
-                nan_mask = torch.isnan(column_data)
+
+        # Since we are not changing the number of items in each column, it's
+        # faster to just clone the values, while reusing the same offset
+        # object.
+        if isinstance(feat, Tensor):
+            if get_na_mask(feat).any():
+                feat = feat.clone()
+            else:
+                return feat
```
Comment (weihua916) [inline torch_frame/nn/encoder/stype_encoder.py:193]: super nit: shall we assert when `feat` is of `dict` type? I know it's covered in `self.na_strategy is None`, but a reader may wonder why we don't cover `isinstance(feat, dict)` case
